qep-335-adopt-wasm32-emscripten-target.md

```diff
@@ -0,0 +1,76 @@
+# QGIS Enhancement: Adopt `wasm32-emscripten` as a build target for QGIS
+
+**Date** 2025/03/10
+
+**Author** Michael Schmuki (@boardend )
+
+**Contact** michael at opengis dot ch
+
+**Version** QGIS 3.44
+
+# Summary
+
+[qgis-js](https://github.com/qgis/qgis-js) has demonstrated that it is possible to cross-compile QGIS core and it's dependencies with [Emscripten](https://emscripten.org/) to [WebAssembly](https://webassembly.org/) in order to use it on the web platform as well as other JavaScript runtimes like Node and Deno.
+
+To achieve this, [some patching](https://github.com/qgis/qgis-js/tree/main/build/vcpkg-ports/qgis-qt6) of QGIS was needed:
+- Some small patches to disable certain features that are not (yet) supported
+- Tweaking of the CMake structure/scripts
+- Some heavy, horizontal patches, for example, the exclusion of the `QgsAuthManager`
+
+Especially the horizontal patches prevent qgis-js from following the QGIS roadmap and keeping it up to date to the latest QGIS version. With this proposal we want to bring those patches upstream to the QGIS project and adopt `wasm32-emscripten` as a supported build target for QGIS.
+
+The primary goal of qgis-js is to port QGIS core to WebAssembly for rendering QGIS projects on the client side, serving as a replacement for QGIS Server. It also includes optional OpenLayers integration for seamless embedding into web applications. While qgis-js's API surface is currently limited, we aim to expand it in the future.
+
+Additionally, compiling QGIS to WebAssembly opens new possibilities, such as porting QGIS processing, bringing PyQGIS to the browser, and building full-screen QGIS apps using Qt Classic or Qt Quick (e.g., porting QField). Although this is outside the scope of qgis-js, upstreaming this work will provide a foundation for developers to explore these opportunities.
+
+## Proposed Solution
+
+The proposed solution involves the following steps:
+1. Rebase all patches from qgis-js and integrate them into the main QGIS repository for use by qgis-js and other projects.
+2. Update qgis-js to the latest QGIS version and ensure future compatibility and updates.
+
+## Deliverables
+
+1. Integration of qgis-js patches into the main QGIS repository on master.
+2. Adopting `wasm32-emscripten` as a build target in CMake/vcpkg
+3. Continuous Integration test on Github Actions to ensure Emscripten compatibility in the future.
+4. A new major version of qgis-js based on QGIS 3.44
+
+### Affected Files
+
+1. CMake build scripts
+2. Various source files (e.g. `#ifdef EMSCRIPTEN`)
```

Comment:
I'm feeling uneasy about adding #ifdef according to a build target all over the code base. 

Could we have instead have #ifdef related to meaningfull build option (like WITH_3D, WITH_ANALYSIS, ... or WITH_AUTH) ?

Reply:
Completely agree, will change that to `WITH_EMSCRIPTEN`

Reply:
I don't think a new build option `WITH_EMSCRIPTEN` is needed (and also no `EMSCRIPTEN`) there is already https://doc.qt.io/qt-5/qtglobal.html#Q_OS_WASM which can be used for features that are platform dependant. We already have this for osx or windows specific features / platform integration.
For things that could possibly also be interesting to disable on other platforms that can be seen as feature flags (just like `WITH_AUTH`) I'd also prefer new feature flags over hardcoded platform specific logic.
In short: both make sense depending on the use case -- and the platform specific definition is already provided through Qt.

Reply:
@troopa81 & @m-kuhn I've just revised the "Affected Files" section. Give me a :+1: if you think it fits like this to resolve this conversation.

Reply:
There is still mention of the `#ifdef EMSCRIPTEN` without too much of insight of what it actually is and where you plan to add it. Could you be more specific, meaning you plan to add some feature related flags (WITH_AUTH for instance) and Q_OS_WASM when we have to disable thing deep in the code?

Reply:
I agree with the recent modifications